### ISSUE TITLE
Help understanding Regex

### ISSUE THREAD

**Issue by Greysys** [aj3423/SpamBlocker#124]
I am playing around with regex to make filtering more useful to me. My use case is blocking calls from EVERY COUNTRY, except those starting with Portugal's country code (+351).

As I understand I can achieve this by using the regex (?:351)?.* (to allow Portuguese numbers), however when I do this it actually allows ALL numbers through, even those from other countries, as if the code was simply .* 

The only workaround here I have found so far is to also add the most common beginning numbers in portuguese phone numbers (e.g. 91,92,96), which leaves me with rules such as (?:351)?96.*, (?:351)?91.*, (?:351)?92.*, and so on.

My question is, is this the only way? Do I need to create a rule for every possible beginning number coming from Portugal and authorize them individually, or is the another, simpler way to do this?

kind regards.

**Comment by aj3423** [comment]
The explanation of your regex by AI:
`(?:351)?.*`
 - Optionally starts with the number "351".
 - Followed by any number of characters
 
It's basically equivalent to `.*`, and the `(?:351)?96.*` == `96.*`

I'd suggest using two simple regex:
 - `.*` priority: 0,   it blocks all numbers
 - `351.*` priority:1, it allows numbers start with 351, with higher priority, so it overrides the previous rule. The leading `+` is ignored.

**Comment by Greysys** [comment]
> .* priority: 0, it blocks all numbers
> 351.* priority:1, it allows numbers start with 351, with higher priority, so it overrides the previous rule. The leading + is ignored.

I have attempted this strategy, but what happens is the call is blocked by the app under the .* rule, because it does not read the +351 country code when I receive a call, it only reads the beginning 91 numbers. In other words, it seems that either the app or my phone aren't recognising the +351 country code by default.

**Comment by aj3423** [comment]
Then it's not an international call, it's just a domestic call, there's no +351 on the pop-up dailer screen, right? The app wouldn't know if there is no +351 on the dialer screen. 

I think you need to find a list of all leading numbers, and add a rule like
`(96|91|92).*`

**Comment by Greysys** [comment]
That's correct. I am currently in Portugal but I usually live abroad and I sometimes receive spam calls from wherever in the world, and the country code is relevant in those instances. I always want to be able to receive a call from Portugal when I'm abroad. 

So your suggestion for my use case would be having the following rules:

1)    `.*`, priority 0 to block all numbers
2)    `(96|91|92).*`, priority 1 to allow all numbers starting with 96,92...

Is that correct? In this instance, how would the country code be checked for? Also, if I already have to add all number prefixes for Portugal anyway, would it make sense to condense everything into one rule? Something like `(?:351)?(96|91|92).*`?

**Comment by aj3423** [comment]
I don't know much about the carrier things, but when the carrier send the call information to your phone, the country code is not always included. In most cases for domestic calls, there is no country code. 

If you don't see +351 on the dialer screen, then it's not included in this call information, and this app wouldn't be able to know it. 

Since you are currently testing it in Portugal, it will always be domestic call and no leading +351. But when you are abroad, there will be +351 because that's international call, and these rules would work.

I think when you are abroad, the whitelist rule `351.*` would be enough, no need for the 91,92,96...

**Comment by Greysys** [comment]
You're right, it makes sense when I'm abroad. Thanks, I will do as you recommend. 

One last question that may not be so relevant: does the app also include SHAKEN attestation, or only STIR? I know you advertise STIR only, but I thought it would be worth asking. Thanks.

**Comment by aj3423** [comment]
This document https://developer.android.com/develop/connectivity/telecom/dialer-app/prevent-spoofing 

claims:

> STIR/SHAKEN verdict

and

> which surfaces the STIR/SHAKEN verdict from the carrier verification mechanism for the STIR/SHAKEN protocol

Seems shaken is also supported. Under the hood, Android just give you a result of the attestation, either of: 

Valid, Invalid or Unidentified.

This result comes along with the call information, I think it depends on the carrier, if the carrier supports both STIR/SHAKEN then the result represents both. For countries(carriers) that only supports STIR, the result only represent STIR. For others, it's always Unidentified.
